### PR TITLE
Metal setup merge

### PR DESCRIPTION
**Pull request by TimSylvester**
Merging back the result of merging main into metal-setup.

### CODE REVIEW

File: include/mbgl/shaders/mtl/common.hpp
Comment (louwers) [inline include/mbgl/shaders/mtl/common.hpp:3]: In my opinion it is not worth a saving few extra keystrokes for introducing a macro here.

**Review by louwers — CHANGES_REQUESTED** [review]
Suggestion: let's add a line to `ios-ci.yml` that compiles the toolkit with the `//:metal_renderer` flag, so we ensure that it at least continues to compile without errors.

File: include/mbgl/shaders/mtl/common.hpp
Comment (TimSylvester) [reply include/mbgl/shaders/mtl/common.hpp:3]: I added this to avoid needing to have the structure name listed twice, once directly and once as a string, and ensure they stay identical.
I was trying to figure out how to make `MLN_STRINGIZE` ensure that the given symbol is valid, but couldn't make it work.

**Comment by louwers** [comment]
🙏

**Review by stefankarschti — APPROVED** [review]

File: include/mbgl/mtl/renderer_backend.hpp
Comment (louwers) [inline include/mbgl/mtl/renderer_backend.hpp:27]: Should be 'Metal state'
```suggestion
    /// Called prior to rendering to update the internally assumed Metal state.
```

File: include/mbgl/mtl/renderer_backend.hpp
Comment (louwers) [inline include/mbgl/mtl/renderer_backend.hpp:45]: Here too?
```suggestion
    /// Tells the renderer that Metal state has already been set by the windowing toolkit.
```

File: include/mbgl/mtl/renderer_backend.hpp
Comment (louwers) [inline include/mbgl/mtl/renderer_backend.hpp:55]:
```suggestion
    /// Triggers an Metal state update if the internal assumed state doesn't
```

File: src/mbgl/mtl/buffer_resource.cpp
Comment (louwers) [inline src/mbgl/mtl/buffer_resource.cpp:35]: Stray comment.

File: include/mbgl/shaders/mtl/common.hpp
Comment (louwers) [reply include/mbgl/shaders/mtl/common.hpp:3]: How about something like this: https://wandbox.org/permlink/GUq6XSaVVzETU4A4

File: include/mbgl/mtl/buffer_resource.hpp
Comment (louwers) [inline include/mbgl/mtl/buffer_resource.hpp:16]: Is the `usage` parameter obvious to someone familiar with the Metal API?

Otherwise, consider documenting this parameter.

File: include/mbgl/mtl/context.hpp
Comment (louwers) [inline include/mbgl/mtl/context.hpp:52]: I am assuming this comment block will be uncommented in a future PR.

File: include/mbgl/mtl/context.hpp
Comment (louwers) [inline include/mbgl/mtl/context.hpp:86]:
```suggestion
    // Only call this while the Metal context is exclusive to this thread.
```

File: include/mbgl/mtl/texture2d.hpp
Comment (louwers) [inline include/mbgl/mtl/texture2d.hpp:23]: The `public` keyword is used three times in a row in this class, and `private` twice in a row.

File: include/mbgl/shaders/mtl/common.hpp
Comment (TimSylvester) [reply include/mbgl/shaders/mtl/common.hpp:3]: Clever, but that's going to create a lot of changes.  I think it should wait until #1618 is merged, and Stefan's work with string indexing should make some of those obsolete, anyway.

File: include/mbgl/shaders/layer_ubo.hpp
Comment (louwers) [inline include/mbgl/shaders/layer_ubo.hpp:8]: I suggest we get rid of this if we decide we don't need the other macro.

File: include/mbgl/mtl/texture2d.hpp
Comment (TimSylvester) [reply include/mbgl/mtl/texture2d.hpp:23]: I believe @alexcristici's render target work makes major changes to `Texture2D`.

File: include/mbgl/mtl/context.hpp
Comment (TimSylvester) [reply include/mbgl/mtl/context.hpp:52]: Yes, the target/framebuffer stuff should be implemented shortly.  I've removed some of that which definitely won't be replaced.

File: include/mbgl/mtl/buffer_resource.hpp
Comment (TimSylvester) [reply include/mbgl/mtl/buffer_resource.hpp:16]: Good catch, that's the wrong type anyway.

File: platform/ios/scripts/bazel-generate-plists.sh
Comment (louwers) [inline platform/ios/scripts/bazel-generate-plists.sh:85]:
```suggestion
BUILD_MODE = "bazel"
```

File: include/mbgl/shaders/mtl/common.hpp
Comment (louwers) [reply include/mbgl/shaders/mtl/common.hpp:3]: OK sure!

File: platform/ios/src/MLNMapView+Metal.mm
Comment (louwers) [inline platform/ios/src/MLNMapView+Metal.mm:117]: Is this intentional?

File: platform/ios/src/MLNMapView+Metal.mm
Comment (louwers) [inline platform/ios/src/MLNMapView+Metal.mm:101]: Is this even still used somewhere?

File: platform/ios/src/MLNMapView+Metal.mm
Comment (louwers) [inline platform/ios/src/MLNMapView+Metal.mm:140]: This comment should be removed, because I doubt it applies for Metal.

File: platform/ios/src/MLNMapView+Metal.mm
Comment (louwers) [inline platform/ios/src/MLNMapView+Metal.mm:155]: Can be removed?

File: src/mbgl/mtl/context.cpp
Comment (louwers) [inline src/mbgl/mtl/context.cpp:36]: Probably TODO?

File: src/mbgl/renderer/layers/circle_layer_tweaker.cpp
Comment (louwers) [inline src/mbgl/renderer/layers/circle_layer_tweaker.cpp:116]: I am not sure what purpose this usage of `MLN_STRINGIZE` serves here. Can we get rid of it?

**Review by louwers — APPROVED** [review]

File: platform/ios/src/MLNMapView+Metal.mm
Comment (TimSylvester) [reply platform/ios/src/MLNMapView+Metal.mm:101]: I don't see any references to it, even in legacy/OpenGL

File: platform/ios/src/MLNMapView+Metal.mm
Comment (TimSylvester) [reply platform/ios/src/MLNMapView+Metal.mm:117]: Doesn't look familiar, @alexcristici?  I think it's just left over from basing the Metal impl on OpenGL, and can be dropped, as can every trace of global state management other than the Context.

Comment (alexcristici) [reply platform/ios/src/MLNMapView+Metal.mm:117]: Yes, we should remove it. 👍

File: src/mbgl/mtl/context.cpp
Comment (alexcristici) [inline src/mbgl/mtl/context.cpp:84]: These 3 functions are available starting with iOS 16, so we should remove them. Or better we should remove the entire `options` block, I don't see where it is used.

**Review by alexcristici — APPROVED** [review]

File: src/mbgl/mtl/context.cpp
Comment (TimSylvester) [reply src/mbgl/mtl/context.cpp:84]: Good catch on it not being used, that's just a bug.  We may want to make the optimization level available to improve initial load time, we just need to make it conditional on the target.

File: include/mbgl/gfx/vertex_attribute.hpp
Comment (mwilsnd) [inline include/mbgl/gfx/vertex_attribute.hpp:114]: Need a static cast here, narrowing `value` from `size_t` to `int`

**Comment by TimSylvester** [comment]
Looks like there's an issue with depth between extrusion and symbols here.

File: src/mbgl/mtl/layer_group.cpp
Comment (mwilsnd) [inline src/mbgl/mtl/layer_group.cpp:31]: `drawableGL` should probably get renamed here to avoid confusion

**Review by mwilsnd — CHANGES_REQUESTED** [review]

File: src/mbgl/mtl/texture2d.cpp
Comment (stefankarschti) [inline src/mbgl/mtl/texture2d.cpp:84]: `noexcept` method throws

**Review by stefankarschti — CHANGES_REQUESTED** [review]

File: platform/ios/src/MLNMapView+Metal.mm
Comment (TimSylvester) [reply platform/ios/src/MLNMapView+Metal.mm:154]: The only other combined option available for iOS doesn't have depth.

```
    MTLPixelFormatDepth24Unorm_Stencil8  API_AVAILABLE(macos(10.11), macCatalyst(13.0)) API_UNAVAILABLE(ios) = 255,
    MTLPixelFormatDepth32Float_Stencil8  API_AVAILABLE(macos(10.11), ios(9.0)) = 260,
    MTLPixelFormatX32_Stencil8  API_AVAILABLE(macos(10.12), ios(10.0)) = 261,
    MTLPixelFormatX24_Stencil8  API_AVAILABLE(macos(10.12), macCatalyst(13.0)) API_UNAVAILABLE(ios) = 262,

```

File: platform/ios/src/MLNMapView+Metal.mm
Comment (TimSylvester) [reply platform/ios/src/MLNMapView+Metal.mm:162]: It looks like `setPresentsWithTransaction` is expected to only be called after `createView`, but it's easy enough to change that.

File: src/mbgl/mtl/context.cpp
Comment (TimSylvester) [reply src/mbgl/mtl/context.cpp]: Not sure what you mean by the shader header.  I don't see a benefit to putting it in `mbgl/gfx/shader.hpp`, `mbgl/shaders/mtl/shader_program.hpp`, etc..  I've just set it to `...2_1` inline for now, the latest version available on iOS 12.0, and made a few minor changes to make our shaders compatible with that version.

**Review by stefankarschti — APPROVED** [review]

File: src/mbgl/mtl/context.cpp
Comment (mwilsnd) [reply src/mbgl/mtl/context.cpp]: I was thinking have it in the generated headers for each shader as we've already got other metadata stored there. Having it in a single place and forcing all shaders to conform is fine too.

**Review by mwilsnd — APPROVED** [review]

**Pull request merged**
